File: VaultGuard.Android/app/src/main/java/com/vaultguard/app/domain/VaultCrypto.kt
package com.vaultguard.app.domain

import android.util.Base64
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec
import javax.inject.Inject
import javax.inject.Singleton

/**
 * On-device encryption for LOCAL connection mode. Byte-for-byte compatible with the desktop
 * `VaultGuard.Crypto` scheme:
 *  - Key derivation: PBKDF2-HMAC-SHA256, 600,000 iterations, 32-byte (256-bit) key (OWASP 2024).
 *  - Cipher: AES-256-GCM, 12-byte nonce, 16-byte auth tag, no associated data.
 *  - Container: the same three components WPF stores in `EncryptedPasswordData`
 *    (Nonce, EncryptedPassword/ciphertext, AuthenticationTag), here joined as
 *    `base64(nonce):base64(ciphertext):base64(tag)`.
 *
 * This means a value encrypted here maps directly onto WPF's EncryptedPassword/Nonce/PasswordAuthTag
 * columns and vice-versa (given the same derived key). In API mode the server does the crypto instead.
 */
@Singleton
class VaultCrypto @Inject constructor() {

    fun newSalt(): ByteArray = ByteArray(SALT_BYTES).also { SecureRandom().nextBytes(it) }

    fun deriveKey(masterPassword: String, salt: ByteArray): ByteArray {
        val spec = PBEKeySpec(masterPassword.toCharArray(), salt, ITERATIONS, KEY_BITS)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        return factory.generateSecret(spec).encoded
    }

    fun encrypt(plaintext: String, key: ByteArray): String {
        val nonce = ByteArray(NONCE_BYTES).also { SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance(TRANSFORM)
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BITS, nonce))
        // Java's GCM output is ciphertext || tag(16); split them to match WPF's separate fields.
        val ctWithTag = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))
        val ct = ctWithTag.copyOfRange(0, ctWithTag.size - TAG_BYTES)
        val tag = ctWithTag.copyOfRange(ctWithTag.size - TAG_BYTES, ctWithTag.size)
        return "${b64(nonce)}:${b64(ct)}:${b64(tag)}"
    }

    fun decrypt(encoded: String, key: ByteArray): String {
        val parts = encoded.split(':')
        require(parts.size == 3) { "Unexpected encrypted value format" }
        val nonce = fromBase64(parts[0])
        val ct = fromBase64(parts[1])
        val tag = fromBase64(parts[2])
        val cipher = Cipher.getInstance(TRANSFORM)
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BITS, nonce))
        return String(cipher.doFinal(ct + tag), Charsets.UTF_8)
    }

    fun toBase64(bytes: ByteArray): String = b64(bytes)
    fun fromBase64(s: String): ByteArray = Base64.decode(s, Base64.NO_WRAP)
    private fun b64(bytes: ByteArray): String = Base64.encodeToString(bytes, Base64.NO_WRAP)

    private companion object {
        const val ITERATIONS = 600_000
        const val KEY_BITS = 256
        const val SALT_BYTES = 16
        const val NONCE_BYTES = 12
        const val TAG_BYTES = 16
        const val TAG_BITS = 128
        const val TRANSFORM = "AES/GCM/NoPadding"
    }
}

File: VaultGuard.iOS.Swift/VaultGuard/Data/VaultCrypto.swift
// VaultCrypto.swift — on-device encryption for LOCAL mode, byte-for-byte compatible with the desktop
// VaultGuard.Crypto scheme:
//  - Key derivation: PBKDF2-HMAC-SHA256, 600,000 iterations, 32-byte key (CommonCrypto).
//  - Cipher: AES-256-GCM, 12-byte nonce, 16-byte tag, no associated data (CryptoKit).
//  - Container: the three WPF EncryptedPasswordData components joined as
//    base64(nonce):base64(ciphertext):base64(tag) — matching Nonce / EncryptedPassword / AuthenticationTag.
import Foundation
import CryptoKit
import CommonCrypto

struct VaultCrypto {
    private let iterations: UInt32 = 600_000

    func newSalt() -> Data {
        var bytes = [UInt8](repeating: 0, count: 16)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes)
    }

    func deriveKey(masterPassword: String, salt: Data) -> SymmetricKey {
        var derived = [UInt8](repeating: 0, count: 32)
        let pw = Array(masterPassword.utf8)
        let saltBytes = [UInt8](salt)
        _ = pw.withUnsafeBufferPointer { pwPtr in
            saltBytes.withUnsafeBufferPointer { saltPtr in
                CCKeyDerivationPBKDF(
                    CCPBKDFAlgorithm(kCCPBKDF2),
                    pwPtr.baseAddress, pw.count,
                    saltPtr.baseAddress, saltBytes.count,
                    CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                    iterations,
                    &derived, derived.count
                )
            }
        }
        return SymmetricKey(data: Data(derived))
    }

    /// Returns base64(nonce):base64(ciphertext):base64(tag), matching WPF's EncryptedPasswordData fields.
    func encrypt(_ plaintext: String, key: SymmetricKey) throws -> String {
        let sealed = try AES.GCM.seal(Data(plaintext.utf8), using: key)
        let nonce = Data(sealed.nonce)
        return "\(nonce.base64EncodedString()):\(sealed.ciphertext.base64EncodedString()):\(sealed.tag.base64EncodedString())"
    }

    func decrypt(_ encoded: String, key: SymmetricKey) throws -> String {
        let parts = encoded.split(separator: ":", omittingEmptySubsequences: false).map(String.init)
        guard parts.count == 3,
              let nonceData = Data(base64Encoded: parts[0]),
              let ct = Data(base64Encoded: parts[1]),
              let tag = Data(base64Encoded: parts[2]) else { throw CryptoError.badData }
        let box = try AES.GCM.SealedBox(nonce: try AES.GCM.Nonce(data: nonceData), ciphertext: ct, tag: tag)
        let opened = try AES.GCM.open(box, using: key)
        guard let s = String(data: opened, encoding: .utf8) else { throw CryptoError.badData }
        return s
    }

    enum CryptoError: Error { case encryptFailed, badData }
}
